dask/dataframe/shuffle.py

```diff
@@ -835,7 +835,10 @@ def partitioning_index(df, npartitions, cast_dtype=None):
         # Fixme: astype raises with strings in numeric columns, but raising
         # here might be very noisy
         df = df.astype(cast_dtype, errors="ignore")
-    return hash_object_dispatch(df, index=False) % int(npartitions)
+    res = hash_object_dispatch(df, index=False) % int(npartitions)
+    # Note: Use a signed integer since pandas is more efficient at handling
+    # this since there is not always a fastpath for uints
+    return res.astype(np.min_scalar_type(-(npartitions - 1)))
```

Comment:
We already do something similar in [shuffle_group](https://github.com/dask/dask/blob/0fac4782129d77ce108ecdd6a98aa5dcedb437ae/dask/dataframe/shuffle.py#L977) for task-based shuffling, so it shouldn't be a problem for cudf (which should support the necessary integer dtypes).